File: apollo-api/src/main/java/com/apollographql/apollo/api/ResponseReader.java
package com.apollographql.apollo.api;

import java.util.List;

/*
 * ResponseReader is an abstraction for reading GraphQL fields.
 */
public interface ResponseReader {

  String readString(ResponseField field);

  Integer readInt(ResponseField field);

  Long readLong(ResponseField field);

  Double readDouble(ResponseField field);

  Boolean readBoolean(ResponseField field);

  <T> T readObject(ResponseField field, ObjectReader<T> objectReader);

  <T> List<T> readList(ResponseField field, ListReader<T> listReader);

  <T> T readCustomType(ResponseField.CustomTypeField field);

  <T> T readConditional(ResponseField field, ConditionalTypeReader<T> conditionalTypeReader);

  interface ObjectReader<T> {
    T read(ResponseReader reader);
  }

  interface ListReader<T> {
    T read(ListItemReader reader);
  }

  interface ConditionalTypeReader<T> {
    T read(String conditionalType, ResponseReader reader);
  }

  interface ListItemReader {

    String readString();

    Integer readInt();

    Long readLong();

    Double readDouble();

    Boolean readBoolean();

    <T> T readCustomType(ScalarType scalarType);

    <T> T readObject(ObjectReader<T> objectReader);

    <T> List<T> readList(ListReader<T> listReader);
  }
}

File: apollo-api/src/main/java/com/apollographql/apollo/api/ResponseReader.kt
package com.apollographql.apollo.api

/*
 * ResponseReader is an abstraction for reading GraphQL fields.
 */
interface ResponseReader {

  fun readString(field: ResponseField): String

  fun readInt(field: ResponseField): Int?

  fun readLong(field: ResponseField): Long?

  fun readDouble(field: ResponseField): Double?

  fun readBoolean(field: ResponseField): Boolean?

  fun <T> readObject(field: ResponseField, objectReader: ObjectReader<T>): T

  fun <T> readList(field: ResponseField, listReader: ListReader<T>): List<T>

  fun <T> readCustomType(field: ResponseField.CustomTypeField): T

  fun <T> readConditional(field: ResponseField, conditionalTypeReader: ConditionalTypeReader<T>): T

  interface ObjectReader<T> {
    fun read(reader: ResponseReader): T
  }

  interface ListReader<T> {
    fun read(reader: ListItemReader): T
  }

  interface ConditionalTypeReader<T> {
    fun read(conditionalType: String, reader: ResponseReader): T
  }

  interface ListItemReader {

    fun readString(): String

    fun readInt(): Int?

    fun readLong(): Long?

    fun readDouble(): Double?

    fun readBoolean(): Boolean?

    fun <T> readCustomType(scalarType: ScalarType): T

    fun <T> readObject(objectReader: ObjectReader<T>): T

    fun <T> readList(listReader: ListReader<T>): List<T>
  }
}
